File: 345_Reverse_Vowels_of_a_String.cpp
# include <iostream>
# include <ctype.h>
# include <string>
using namespace std;

class Solution {
public:
    string reverseVowels(string s) {
        int i = 0, j = s.length()-1;
        while (i < j) {
            while (!isValid(s[i]) && i < j) i++;
            while (!isValid(s[j]) && i < j) j--;
            if (i < j) swap(s[i++], s[j--]);
        }
        return s;
    }

    bool isValid(char c) {
        c = tolower(c);
        char a[] = "aeiou";
        for (int i = 0; i < 6; i++)
            if (a[i] == c) return true;
        return false;
    }
};

File: 345_Reverse_Vowels_of_a_String.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        arr = list(s)
        i, j = 0, len(s)-1
        ch = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'}
        while i < j:
            while arr[i] not in ch and i < j:
                i += 1
            while arr[j] not in ch and i < j:
                j -= 1
            if i < j:
                arr[i], arr[j] = arr[j], arr[i]
                i += 1
                j -= 1
        return ''.join(arr)
